AK/URL.cpp

```diff
@@ -467,4 +488,97 @@ String URL::basename() const
+    VERIFY_NOT_REACHED();
+}
+
+// https://url.spec.whatwg.org/#c0-control-percent-encode-set
+constexpr bool code_point_is_in_percent_encode_set(u32 code_point, URL::PercentEncodeSet set)
+{
+    switch (set) {
+    case URL::PercentEncodeSet::C0Control:
+        return code_point < 0x20 || code_point > 0x7E;
+    case URL::PercentEncodeSet::Fragment:
+        return code_point_is_in_percent_encode_set(code_point, URL::PercentEncodeSet::C0Control) || " \"<>`"sv.contains(code_point);
+    case URL::PercentEncodeSet::Query:
+        return code_point_is_in_percent_encode_set(code_point, URL::PercentEncodeSet::C0Control) || " \"#<>"sv.contains(code_point);
+    case URL::PercentEncodeSet::SpecialQuery:
+        return code_point_is_in_percent_encode_set(code_point, URL::PercentEncodeSet::Query) || code_point == '\'';
+    case URL::PercentEncodeSet::Path:
+        return code_point_is_in_percent_encode_set(code_point, URL::PercentEncodeSet::Query) || "?`{}"sv.contains(code_point);
+    case URL::PercentEncodeSet::Userinfo:
+        return code_point_is_in_percent_encode_set(code_point, URL::PercentEncodeSet::Path) || "/:;=@[\\]^|"sv.contains(code_point);
+    case URL::PercentEncodeSet::Component:
+        return code_point_is_in_percent_encode_set(code_point, URL::PercentEncodeSet::Userinfo) || "$%&+,"sv.contains(code_point);
+    case URL::PercentEncodeSet::ApplicationXWWWFormUrlencoded:
+        return code_point >= 0x7E || !(is_ascii_alphanumeric(code_point) || "!'()~"sv.contains(code_point));
+    case URL::PercentEncodeSet::EncodeURI:
+        // NOTE: This is the same percent encode set that JS encodeURI() uses.
+        // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURI
+        return code_point >= 0x7E || (!is_ascii_alphanumeric(code_point) && !";,/?:@&=+$-_.!~*'()#"sv.contains(code_point));
+    default:
+        VERIFY_NOT_REACHED();
+    }
+}
+
+String URL::percent_encode(u32 code_point, URL::PercentEncodeSet set)
+{
+    if (code_point_is_in_percent_encode_set(code_point, set))
+        return percent_encode_code_point(code_point);
+    StringBuilder builder(4);
+    builder.append_code_point(code_point);
+    return builder.to_string();
+}
+
+String URL::percent_encode(const StringView& input, URL::PercentEncodeSet set)
+{
+    StringBuilder builder;
+    for (auto code_point : Utf8View(input)) {
+        builder.append(percent_encode(code_point, set));
+    }
+    return builder.to_string();
+}
+
+constexpr u8 parse_hex_digit(u8 digit)
+{
+    if (digit >= '0' && digit <= '9')
+        return digit - '0';
+    if (digit >= 'a' && digit <= 'f')
+        return digit - 'a' + 10;
+    if (digit >= 'A' && digit <= 'F')
+        return digit - 'A' + 10;
+    VERIFY_NOT_REACHED();
+}
+
+String URL::percent_decode(const StringView& input)
```

Comment:
I know what you mean, but since probably all usages of this function just assume it to be a `String` and convert it to either a `String` or `StringView` anyway, it does not make much sense at this point.